core/src/main/scala/org/apache/spark/shuffle/PushShuffleSupport.scala

```diff
@@ -0,0 +1,466 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import java.io.File
+import java.net.ConnectException
+import java.nio.ByteBuffer
+import java.util.concurrent.ExecutorService
+
+import scala.collection.mutable.{ArrayBuffer, HashMap, HashSet, Queue}
+
+import com.google.common.base.Throwables
+
+import org.apache.spark.{ShuffleDependency, SparkConf, SparkEnv}
+import org.apache.spark.annotation.Since
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.network.buffer.{FileSegmentManagedBuffer, ManagedBuffer, NioManagedBuffer}
+import org.apache.spark.network.netty.SparkTransportConf
+import org.apache.spark.network.shuffle.BlockFetchingListener
+import org.apache.spark.network.shuffle.ErrorHandler.BlockPushErrorHandler
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.shuffle.PushShuffleSupport._
+import org.apache.spark.storage.{BlockId, BlockManagerId, ShufflePushBlockId}
+import org.apache.spark.util.{ThreadUtils, Utils}
+
+/**
+ * Used for pushing shuffle blocks to remote shuffle services when push shuffle is enabled.
+ * When push shuffle is enabled, it is created after the shuffle writer finishes writing the shuffle
+ * file and initiates the block push process.
+ *
+ * @param dataFile         mapper generated shuffle data file
+ * @param partitionLengths array of shuffle block size so we can tell shuffle block
+ *                         boundaries within the shuffle file
+ * @param dep              shuffle dependency to get shuffle ID and the location of remote shuffle
+ *                         services to push local shuffle blocks
+ * @param partitionId      map index of the shuffle map task
+ * @param conf             spark configuration
+ */
+private[spark] class PushShuffleSupport(
```

Comment:
XXXSupport is usually a trait or abstract class.
Would it be better to make `PushShuffleSupport` an abstract class which `ShuffleWriter` extends?
The `getPartitionLengths` method can be the abstract method in this trait instead of inside `ShuffleWriter`.
All the other class fields in this class can be converted into method arguments for `initiateBlockPush`.
This way, we would still be able to initiate block push via the ShuffleWriter objects.

Reply:
> XXXSupport is usually a trait or abstract class.

I can rename it to be `PushShuffleComponent`

> Would it be better to make PushShuffleSupport an abstract class which ShuffleWriter extends?
The getPartitionLengths method can be the abstract method in this abstract class instead of inside ShuffleWriter

I can add another interface which let's say looks like this and ShuffleWriter can extend it
```
interface PushEnabledShuffleWriter {
 getPartitionLengths()
 initiateBlockPush(...)
}
```
Not sure if the interface is really needed since ShuffleWriter is the only base writer to which we are adding this ability. Maybe we can just add back the `initiateBlockPush` method to ShuffleWriter.

> All the other class fields in this class can be converted into method arguments for initiateBlockPush.
This way, we would still be able to initiate block push via the ShuffleWriter objects instead of relying on instantiating an object of PushShuffleSupport.

With the above interface, ` initiateBlockPush` of `ShuffleWriter` can delegate the push to `PushShuffleComponent`.

Reply:
I have renamed `PushShuffleSupport` to `PushShuffleComponent`. Will wait for more feedback to make anymore changes to this.